README
MY BACKUP README
================
version 2.3

Copyright (C) 2021 António Manuel Dias

contact: ammdias@gmail.com

website: https://ammdias.duckdns.org/downloads

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see http://www.gnu.org/licenses/.


Changes history:

* 2.3: Corrected bug that prevented manual from being displayed.
* 2.2.1: Corrected bug that prevented an 'ok' to be shown after correctly
             parsing a composite profile.
* 2.2: Added option to edit the configuration file
* 2.1: Added option to list backup profiles
       Added option to display configuration file
       Composite profiles can now only be composed of simple profiles,
           thus preventing 'infinite loop' profiles
* 2.0: Complete rewrite in Python.
       Configuration with possible multiple profiles in a separate file.
* 1.0: Original bash script.


ABOUT THE PROGRAM
=================

This program makes easy the backup of a group of directories to another
location or remote directory, with the help of an external utility like
`rsync` or `cp`.  It depends on Python 3.

The following instructions are for Linux systems.  If you intend to use the
program on another system, please refer to the included MANUAL.


USAGE
=====

Initial test of the program
---------------------------

1. Uncompress the program zip file to a convenient place and edit the 
   configuration file, "mybkp_profiles", with your favorite text editor.

   The configuration file is divided in "profiles", each starting with
   the profile name between brackets.  Each profile must have four values:

   * `command`: the command to perform the backup (e.g. `rsync` or `cp`),
     including all the commands' options
   * `base`: the parent directory where the directories to backup reside
     (e.g. your home directory)
   * `backup`: the directory where the backup will reside (e.g. an external
     disk drive or network storage)
   * `directories`: a comma separated list of directories, below the `base`
     directory, to backup
   
   The provided sample file comes with a single profile, named "default":
   
       [default]
       command: echo cp --archive --update --verbose --strip-trailing-slashes
       base: /home/USER
       backup: /home/USER/tmp
       directories: Documents, Desktop
   
   This profile, when used, will backup the "Documents" and "Desktop"
   directories under the user's home directory ("/home/USER"), using
   the command

       echo cp --archive --update --verbose --strip-trailing-slashes
   
   (Actually, it will do nothing but print the commands to the terminal, as
    that is all the `echo` program does :)
   
2. To try it out, open a terminal and change to the directory where you
   uncompressed the program.  Then, make sure the program is executable
   and execute it, passing the configuration file and profile as argument:
   
       $ chmod +x mybkp.py
       $ ./mybkp.py --config mybkp_profiles default
   
   It should output something like this:
   
       Parsing profile: default... ok.
   
       Processing profile: default
       Backing up directory Documents ...
       cp --archive --update --verbose --strip-trailing-slashes \
          home/USER/Documents/ /home/USER/tmp
   
       Backing up directory Desktop ...
       cp --archive --update --verbose --strip-trailing-slashes \
          /home/USER/Desktop/ /home/USER/tmp
   
3. If you want to really test it, remove the `echo` from the `command` value,
   replace USER with your username and execute the program again.  Now, it
   should actually copy the "Documents" and "Desktop" directories into the
   "tmp" directory.
   
   (Of course, you can -- probably should -- change the "Documents", "Desktop"
    and "tmp" directories to any others of your choosing).
   
   If you wish to check the commands that will be issued by the program,
   instead of actually execute them, use the `--no_act` option:
   
       $ ./mybkp.py default --config mybkp_profiles --no_act


Installation and actual basic usage
-----------------------------------

The following instructions describe the installation process for basic usage
by a single user in a Linux environment.  For more advanced instructions,
including MS Windows usage, please refer to the user manual in the file
"MANUAL" or "MANUAL.pdf".

1. Open a terminal and change to the directory where you uncompressed the
   program.  Execute the local installation script:

       $ bash local_install.sh

   This will copy the program to the current user's "~/.local/lib" hidden
   directory, create a symbolic link to the program in "~/.local/bin",
   which should be in the user's PATH, and copy the configuration file to
   "~/.config/mybkp_profiles".

2. To check that the program is working, open *another* terminal and type:

       $ mybkp --version

   This should print the program's name and version.  If not, check if
   "~/.local/bin" is in the PATH:

       $ echo $PATH

   Also, check that the program was copied to the locations mentioned above
   and that the symbolic link was created.

3. Now edit the configuration file in "~/.config/mybkp_profiles" with a text
   editor, changing it to your particular needs.
   
   Start with something really simple, as making the backup of two or three
   directories to an external drive.  Make sure of including the full paths
   to the base and backup directories.

4. Run the program in "dry run" mode and check that the copy commands are all
   correct:

       $ mybkp default --no_act

   (notice that when the configuration file is in the correct place, you don't
    need to include it as an option)

5. Finally, run the program and check that the backup was correctly made:

       $ mybkp default

6. To restore the backup (copy the files from the backup location to their
   original place), use the restore option:

       $ mybkp default --restore

7. Read the included manual to learn how to create configuration files with
   multiple profiles and different backup strategies.  For simple command
   line help you may type:

       $ mybkp --help




CHANGES.md
My Backup
=========

(C) 2015 António Manuel Dias

Changes list:

    * 3.1  New install system.
           Added uninstall option.
           Updated README and MANUAL.

    * 3.0: Code refactoring

    * 2.6: Removed creation destination directories because it was unsafe.
           Added confirmation step before starting backup process.

    * 2.5: Verify that source and destination directory are different
           Create destination directory if not exists

    * 2.4: Changed installation scripts.
           Changed README and MANUAL accordingly.

    * 2.3: Removed PDF documents. Added HTML manual and option to show it.

    * 2.2.1: Corrected bug that prevented an 'ok' to be shown after correctly
             parsing a composite profile.

    * 2.2: Added option to edit the configuration file.

    * 2.1: Added option to list backup profiles.
           Added option to display configuration file.
           Composite profiles can now only be composed of simple profiles,
           thus preventing 'infinite loop' profiles.

    * 2.0: Complete rewrite in Python.
           Configuration with possible multiple profiles in separate file.
    
    * 1.0: Original bash script.


cli.py
"""My Backup - Command Line Interface

Backup a group of directories from a base location using an external program
(C) 2015 António Manuel Dias

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

__version__ = '3.1'
__date__ = '2023-12-14'
__license__ ='GNU General Public License version 3'
__author__ = 'António Manuel Dias <ammdias@gmail.com>'


import sys
import os
import webbrowser
from core import *
from mybkp_text import *


def cli_run(args):
    """Run the program from the command line."""
    # options for displaying general information
    if args.version:
        prt_exit(f'{NAME} {__version__}')

    if args.copyright:
        prt_exit(f'{NAME} {__version__}\n\n{SHORT_COPYRIGHT}\n\n{COPYRIGHT}')

    if args.warranty:
        prt_exit(f'{NAME} {__version__}\n\n{SHORT_COPYRIGHT}\n\n{WARRANTY}')

    if args.uninstall:
        from UNINSTALL import uninstall
        uninstall()

    if args.manual:
        webbrowser.open_new(os.path.join(sys.path[0], 'MANUAL.html'))
        prt_exit(MANUAL)

    if args.include_destination_directories:
        prt_exit(WARNING_IDD, 1)
 
    # get configuration file path
    config_file = find_config_file(args.config)
    if not config_file:
        prt_exit(CONFIG_NOT_FOUND, 1)

    # options for manipulating config file
    if args.print_config:
        print(CONFIG_AT.format(config_file))
        prt_exit(open(config_file, 'r').read())

    if args.edit_config:
        edit_config(config_file)
        prt_exit('')

    # parse configuration file and build backup object
    try:
        valid_profiles = Profiles(config_file)
    except Exception as e:
        prt_exit(PARSING_EXCEPTION.format(e))

    # option to list valid profiles
    if args.list_profiles:
        list_profiles(valid_profiles.profiles)
        return

    # perform intended backup/restore operation
    backup(valid_profiles, args.profiles or ('default',),
           restore=args.restore, dryrun=args.no_act,
           bkpdir=args.backup_directory)


def list_profiles(profiles):
    """List valid configuration profiles."""
    for p in profiles:
        if type(profiles[p]) == list:
            print(f"{p}: {', '.join(profiles[p])}")
        else:
            print(p)


def edit_config(config_file):
    """Edit the configuration file with default editor."""
    if 'VISUAL' in os.environ:
        editor = os.environ['VISUAL']
    elif 'EDITOR' in os.environ:
        editor = os.environ['EDITOR']
    else:
        prt_exit(EDITOR_NOT_FOUND, 1)

    try:
        if subprocess.run((editor, config_file)).returncode:
            prt_exit(EDITOR_FAILED, 1)
    except Exception as e:
        prt_exit(EDITOR_EXCEPTION.format(e), 1)
    except:
        prt_exit(EDITOR_EXCEPTION_UNK, 1)


def backup(valid_profiles, profiles, restore, dryrun, bkpdir):
    """Perform backup on each profile."""
    try:
        for setup in valid_profiles.process_profiles(profiles, restore, bkpdir):
            print(PROFILE_SETUP.format(setup['profile'],
                                       setup['source'], setup['destination'],
                                       ', '.join(setup['directories']),
                                       ', '.join(setup['options'])))
            if not yesno(PROCEED):
                continue
            print()
    
            for dirsetup in process_directories(setup):
                if dryrun:
                    print(f"{' '.join(setup['command'])} "
                          f"\"{dirsetup['source']}\" \"{dirsetup['destination']}\"")
                else:
                    print(PROCESS_DIR.format(dirsetup['directory']))
                    exec_command(setup['command'],
                                 dirsetup['source'], dirsetup['destination'])
                    print()

    except ProfileNotFoundError as e:
        prt_exit(PROFILE_NOT_FOUND.format(e), 1)
    except NestedDirectoriesError:
        prt_exit(NESTED_DIR_EXCEPTION)
    except CommandError:
        prt_exit(CMD_ERROR)


def yesno(question):
    """Gets yes or no answer."""
    while True:
        ans = input(question).strip().lower()
        if ans in YES:
            return True
        if ans in NO:
            return False


def prt_exit(msg='', exit_code=0): 
    """Print message and exit."""
    print(msg, file=sys.stderr if exit_code else sys.stdout)
    sys.exit(exit_code)




mybkp.py
#!/usr/bin/env python3

"""My Backup

Backup a group of directories from a base location using an external program
(C) 2015 António Manuel Dias

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

__version__ = '3.1'
__date__ = '2023-12-14'
__license__ ='GNU General Public License version 3'
__author__ = 'António Manuel Dias <ammdias@gmail.com>'


import sys
import argparse
from cli import cli_run, prt_exit
from mybkp_text import *


def parse_arguments():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(description=f'{NAME} {__version__}: {DESCR}')

    parser.add_argument('profiles', type=str, nargs='*', help=ARG_PROFILES)

    parser.add_argument('-c', '--config', type=str, default='', help=ARG_CONFIG)

    parser.add_argument('-b', '--backup-directory', help=ARG_BACKUP_DIRECTORY)

    parser.add_argument('-e', '--edit-config', action='store_true', help=ARG_EDIT_CONFIG)

    parser.add_argument('-i', '--include_destination_directories', action='store_true',
                        help=ARG_IDD_DEPRECATED)

    parser.add_argument('-l', '--list-profiles', action='store_true', help=ARG_LIST_PROFILES)

    parser.add_argument('-n', '--no-act', action='store_true', help=ARG_NO_ACT)

    parser.add_argument('-p', '--print-config', action='store_true', help=ARG_PRINT_CONFIG)

    parser.add_argument('-r', '--restore', action='store_true', help=ARG_RESTORE)

    parser.add_argument('--copyright', action='store_true', help=ARG_COPYRIGHT)

    parser.add_argument('--warranty', action='store_true', help=ARG_WARRANTY)

    parser.add_argument('--manual', action='store_true', help=ARG_MANUAL)

    parser.add_argument('--version', action='store_true', help=ARG_VERSION)

    parser.add_argument('--uninstall', action='store_true', help=ARG_UNINSTALL)

    return parser.parse_args()

#------------------------------------------------------------------------------
# Program entry point

if __name__ == '__main__':
    try:
        args = parse_arguments()
        cli_run(args)
    except KeyboardInterrupt:
        prt_exit(EXIT_EXCEPTION, 1)
    except Exception as e:
        prt_exit(UNK_EXCEPTION.format(e), 1)


CONFIG.py
'''Configuration file for the My Backup installation script.
'''

DOC = 'Installation script for My Backup.'''
COPYRIGHT_YEAR = '2015'
VERSION = '3.1'
DATE = '2023-12-14'
AUTHOR = 'António Manuel Dias <ammdias@gmail.com>'
LICENSE = '''
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''


# List of file to be copied to installation the directory
FILES = ('mybkp.py', 'mybkp_text.py', 'core.py', 'cli.py',
         'UNINSTALL.py', '__version__', 'CHANGES.md',
         'LICENSE.md', 'MANUAL.md', 'MANUAL.html', 'README.md')

# List of directories to be copied to the installation directory
TREES = ()

# Name of the icon file
ICO_FILE = None

# Name of the desktop entry file (for GUI menus)
DESKTOP_FILE = None

# Files to make executable
EXECS = ('mybkp.py',)

# Name of the application (will be the name of the installation directory)
APP_NAME = 'MyBkp'

# Symbolic links to make: dictionary of 'link name': 'executable name' pairs
LINKS = {'mybkp': 'mybkp.py'}

# List of configuration files
CONFIG_FILES = ('mybkp_profiles',)


INSTALL.py
#!/usr/bin/env python3

"""Installation script for My Backup.
"""

FILES = ('LICENSE.md', 'MANUAL.md', 'MANUAL.html', 'README.md',
         'mybkp.py', 'mybkp_text.py', 'core.py', 'cli.py',
         '__version__')
APP_NAME = 'MyBkp' # will be the name of installation directory
START_SCRIPT = 'mybkp.py'
LINK_NAME = 'mybkp'
CONFIG_FILES = ('mybkp_profiles',)  # must be a tuple or list

__version__ = '3.0'
__date__ = '2021-06-25'
__author__ = 'António Manuel Dias <ammdias@gmail.com>'
__license__ = '''
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import sys
import os.path
import shutil
from itertools import zip_longest

PKG_DIR = sys.path[0]
INSTALL_DIR = os.path.join('~', '.local', 'lib')
BIN_DIR = os.path.join('~', '.local', 'bin')
CONFIG_DIR = os.path.join('~', '.config')


def _quit(msg, exc=None):
    '''Print error message and quit.
    '''
    if exc:
        msg = f'{msg} Reason:\n{exc}'
    print(f'Error: {msg}\n', file=sys.stderr)
    sys.exit(1)


def yesno(question):
    """Get a yes or no answer to a question.
    """
    answer = ''
    while answer not in ('y', 'yes', 'n', 'no'):
        answer = input(question + ' (y/n): ').strip().lower()

    return answer in ('y', 'yes')


# print greeting
print(__doc__)
print('Version:', __version__)
print('Copyright (C)', __date__[:4], __author__)
print(__license__)


# get installation directories
d = input(f'Installation directory [{INSTALL_DIR}]: ').strip()
INSTALL_DIR = os.path.expanduser(os.path.join(d or INSTALL_DIR, APP_NAME))
START_SCRIPT_PATH = os.path.join(INSTALL_DIR, START_SCRIPT)

d = input(f'Start link directory [{BIN_DIR}]: ').strip()
BIN_DIR = os.path.expanduser(d or BIN_DIR)
LINK_PATH = os.path.join(BIN_DIR, LINK_NAME)

if CONFIG_FILES:
    d = input(f'Configuration files directory [{CONFIG_DIR}]: ').strip()
    CONFIG_DIR = os.path.expanduser(d or CONFIG_DIR)
    if len(CONFIG_FILES) > 1:
        CONFIG_DIR = os.path.join(CONFIG_DIR, APP_NAME)

print()


# check if install path exists and clean it if it does
if PKG_DIR == INSTALL_DIR or PKG_DIR == BIN_DIR:
    _quit("Installation directory cannot be the same as package directory.")
try:
    if os.path.exists(INSTALL_DIR):
        try:
            with open(os.path.join(INSTALL_DIR, '__version__'), 'r') as f:
                oldversion = f.read().strip()
        except:
            pass  # corrupted previous installation? just remove it.
        else:
            print(f'Version {__version__} will be installed '
                  f'over existing version {oldversion}.')
            if not yesno('Proceed anyway?'):
                _quit("Installation interrupted by user.")
            print()
        print('Removing old version...')
        shutil.rmtree(INSTALL_DIR)
    if os.path.lexists(LINK_PATH):
        print('Removing old symbolic link to startup script...')
        os.remove(LINK_PATH)
except Exception as e:
    _quit("Could not remove old installation.", e)


# copy files to install directory
try:
    print('Creating installation directory...')
    os.makedirs(INSTALL_DIR)
    print('Copying files:')
    for i in FILES:
        print(f'... {i}')
        shutil.copy2(os.path.join(PKG_DIR, i), INSTALL_DIR) 
except Exception as e:
    _quit("Could not copy files to installation directory.", e)


# make symbolic link to application start script
try:
    print('Creating symbolic link to startup script...')
    os.chmod(START_SCRIPT_PATH, 0o755)
    os.makedirs(BIN_DIR, exist_ok=True)
    os.symlink(START_SCRIPT_PATH, LINK_PATH)
except Exception as e:
    _quit("Could not create symbolic link to application script.", e)

# copy default configuration files, if they don't already exist
if CONFIG_FILES:
    try:
        print('Creating configuration files directory...')
        os.makedirs(CONFIG_DIR, exist_ok=True)
        print('Copying necessary configuration files:')
        for i in CONFIG_FILES:
            if not os.path.exists(os.path.join(CONFIG_DIR, i)):
                print(f'... {i}')
                shutil.copy2(os.path.join(PKG_DIR, i), CONFIG_DIR) 
    except Exception as e:
        _quit("Could not copy configuration files.", e)

print('\nApplication successfully installed.\n')



mybkp_text.py
"""My Backup (text constants)

Backup a group of directories from a base location using an external program
(C) 2015 António Manuel Dias

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

__version__ = '3.1'
__date__ = '2023-12-14'
__license__ ='GNU General Public License version 3'
__author__ = 'António Manuel Dias <ammdias@gmail.com>'


ARG_PROFILES = 'name of the backup profile(s) to execute.'
ARG_CONFIG = 'path to the profiles configuration file.'
ARG_BACKUP_DIRECTORY = 'add this directory to the base backup directory.'
ARG_EDIT_CONFIG = "edit the configuration file using the system's default editor and exit."
ARG_IDD_DEPRECATED = 'This option has been deprecated and should not be used.'
ARG_LIST_PROFILES = 'list valid backup profiles in configuration file and exit.'
ARG_NO_ACT = 'show what files to backup, without actually performing the action.'
ARG_PRINT_CONFIG = 'show configuration file and exit.'
ARG_RESTORE = 'restore directories from backup.'
ARG_COPYRIGHT = 'show copyright information and exit.'
ARG_WARRANTY = 'show warranty information and exit.'
ARG_MANUAL = 'display the manual in a web browser window and exit.'
ARG_VERSION = 'show version information and exit.'
ARG_UNINSTALL = 'uninstall application.'

NAME = "My Backup"
DESCR = "Backup a group of directories."

SHORT_COPYRIGHT = """My Backup
(C) 2015 António Manuel Dias
<https://ammdias.duckdns.org/downloads>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>."""

COPYRIGHT = """From the Preamble of the GNU General Public License:
        
The GNU General Public License is a free, copyleft license for
software and other kinds of works.

The licenses for most software and other practical works are designed
to take away your freedom to share and change the works.  By contrast,
the GNU General Public License is intended to guarantee your freedom to
share and change all versions of a program--to make sure it remains free
software for all its users.  We, the Free Software Foundation, use the
GNU General Public License for most of our software; it applies also to
any other work released this way by its authors.  You can apply it to
your programs, too.

When we speak of free software, we are referring to freedom, not
price.  Our General Public Licenses are designed to make sure that you
have the freedom to distribute copies of free software (and charge for
them if you wish), that you receive source code or can get it if you
want it, that you can change the software or use pieces of it in new
free programs, and that you know you can do these things.

To protect your rights, we need to prevent others from denying you
these rights or asking you to surrender the rights.  Therefore, you have
certain responsibilities if you distribute copies of the software, or if
you modify it: responsibilities to respect the freedom of others.

For example, if you distribute copies of such a program, whether
gratis or for a fee, you must pass on to the recipients the same
freedoms that you received.  You must make sure that they, too, receive
or can get the source code.  And you must show them these terms so they
know their rights.

See the file 'gpl.txt' on the program's directory for more details.
If it's missing please refer to http://www.gnu.org/licenses/gpl.txt
"""

WARRANTY = """From the GNU General Public License:
    
15. Disclaimer of Warranty.

THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY
APPLICABLE LAW.  EXCEPT WHEN OTHERWISE STATED IN WRITING THE
COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS"
WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY
AND PERFORMANCE OF THE PROGRAM IS WITH YOU.  SHOULD THE PROGRAM PROVE
DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR
CORRECTION.

16. Limitation of Liability.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR
CONVEYS THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES
ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT
NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES
SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO
OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY
HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
"""

MANUAL = """The program's manual should open in a new web browser window.
If not, please read the manual in text format (actually *Markdown*) in
the program's installation directory.
"""

WARNING_IDD = """
WARNING: the -i / --include_destination_directories option is
no longer valid.  Instead, it has been substituted by the
"include_destination_directories" option in the "options" argument of each
profile.  Please, read the "Configuration file syntax > Available options"
section of the Manual for more information.
"""

CONFIG_AT = "Configuration file at: {}\n"

PROFILE_SETUP = """\nProcessing profile: {}
    From: {}
    To: {}
    Directories: {}
    Options: {}
"""
PROCEED = "Proceed with this operation (y/n)? "
YES = ('y', 'yes')
NO = ('n', 'no')

PROCESS_DIR = "Processing directory {} ..."

CONFIG_NOT_FOUND = """Configuration file not found.
Please provide the correct path to this file as an option
to the program or use the default locations to place it.
See the program's manual for more information on this subject.
"""

EXIT_EXCEPTION = "\nProcess terminated by user."
UNK_EXCEPTION = "\nUnknown error ocurred:\n{}"

EDITOR_NOT_FOUND = "No default editor is configured on your system."
EDITOR_FAILED = "Editor returned with non-zero status."
EDITOR_EXCEPTION = "Exception ocurred while editing the configuration file:\n-->{}"
EDITOR_EXCEPTION_UNK = "Unknown exception ocurred while editing the configuration file."

PROFILE_NOT_FOUND = "Profile not found: {}"
PARSING_EXCEPTION = "Exception occurred while reading the configuration file:\n--> {}"
CONFIG_EXCEPTION = "Exception ocurred when processing profile configuration:\n--> {}"

NESTED_DIR_EXCEPTION = "Error: destination directory inside source directory."
CMD_ERROR = "Profile command exited with non-zero status."



core.py
"""My Backup - Nuclear classes and functions for My Backup program

Backup a group of directories from a base location using an external program
(C) 2015 António Manuel Dias

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

__version__ = '3.1'
__date__ = '2023-12-14'
__license__ ='GNU General Public License version 3'
__author__ = 'António Manuel Dias <ammdias@gmail.com>'


import sys
import os.path
import subprocess
from configparser import ConfigParser, ExtendedInterpolation
from mybkp_text import PROFILE_NOT_FOUND


DEFAULT_CONFIG_FILES = (
    os.path.expanduser(os.path.join('~', '.mybkp_profiles')),
    os.path.expanduser(os.path.join('~', '.config', 'mybkp_profiles'))
)


def find_config_file(config_file):
    """Return configuration file path."""
    config_files = DEFAULT_CONFIG_FILES
    # read install log to check original config file path
    try:
        for i in open(os.path.join(sys.path[0], 'install.log')).readlines():
            ptype, sep, path = i.strip().partition(':')
            if '' in (ptype, sep, path):
                break                     # invalid log, ignore
            if ptype == 'config_file':
                config_files = (path, *config_files)
                break
    except:
        pass                              # error reading log, ignore

    for i in (config_file,) if config_file else config_files:
        if os.path.lexists(i):
            return i

    return None


def process_directories(setup):
    """Loop over directories on a setup configuration,
       processing each one."""
    for d in setup['directories']:
        src, dest = (os.path.join(setup['source'], d, ''), # '' ensures source is a directory
                     os.path.join(setup['destination'], d)
                     if 'include_destination_directories' in setup['options']
                     else setup['destination'])

        if dest.startswith(src):
            raise NestedDirectoriesError()

        yield {
            'directory': d,
            'source': src,
            'destination': dest
        }


def exec_command(command, source, destination):
    """Execute command on source and destination directories."""
    if subprocess.run((*command, source, destination)).returncode:
        raise CommandError()


class Profiles:
    """Dictionary of backup profiles."""

    def __init__(self, config_file):
        """Dictionary of backup profiles.
           config_file: path to the My Backup configuration file.
        """
        config_parser = ConfigParser(interpolation=ExtendedInterpolation())
        config_parser.read(config_file)

        self.profiles = {}
        comp_profiles = []
        base_opts = set(('command', 'base', 'backup', 'directories'))
        comp_opts = base_opts | set(('options',))
        for p in config_parser.sections():
            profile_opts = set(config_parser[p])
            if profile_opts == base_opts:
                self.profiles[p] = config_parser[p]
                self.profiles[p]['options'] = '' 
            elif profile_opts == comp_opts:
                self.profiles[p] = config_parser[p]
            elif 'profiles' in profile_opts:
                comp_profiles.append(p)

        for p in comp_profiles:
            if len(config_parser[p]) == 1:
                c = []
                for pp in map(str.strip, config_parser[p]['profiles'].split(',')):
                    if pp in self.profiles:
                        c.append(pp)
                if c:
                    self.profiles[p] = c


    def process_profiles(self, profiles, restore, bkpdir):
        """Loop over profiles yielding a setup configuration for each one."""
        for profile in profiles:
            p = self.profiles.get(profile, None)
            if not p:
                raise ProfileNotFoundError(profile)

            if type(p) == list:
                profiles += p
                continue

            command = p['command'].split()
            src, dest = p['base'], p['backup']
            directories = list(map(str.strip, p['directories'].split(',')))
            options = list(map(str.strip, p['options'].split(',')))

            if bkpdir:
                dest = os.path.join(dest, bkpdir)

            if restore:
                src, dest = dest, src

            yield {
                'profile': profile,
                'command': command,
                'source': src,
                'destination': dest,
                'directories': directories,
                'options': options
            }


class ProfileNotFoundError(Exception):
    """Raised by an operation on a non-existing profile."""

class NestedDirectoriesError(Exception):
    """Raised when destination directory is inside source directory."""

class CommandError(Exception):
    """Raised when processing command returns non-zero."""

